File: backend/src/models/okr.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ─── Domain Models ───

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrObjective {
    pub workspace_id: Uuid,
    pub id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub objective_type: String,
    pub cycle: String,
    pub status: String,
    pub owner_id: Uuid,
    pub team_id: Option<Uuid>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_ids: Option<Vec<Uuid>>,
    pub parent_objective_id: Option<Uuid>,
    pub progress: f32,
    pub confidence: f32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    // Enriched fields
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key_results: Option<Vec<OkrKeyResult>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub owner_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub children: Option<Vec<OkrObjectiveChild>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrObjectiveChild {
    pub parent_objective_id: Uuid,
    pub child_objective_id: Uuid,
    pub workspace_id: Uuid,
    pub child_title: String,
    pub child_progress: f32,
    pub child_confidence: f32,
    pub child_team_id: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrObjectiveByCycle {
    pub workspace_id: Uuid,
    pub cycle: String,
    pub objective_id: Uuid,
    pub title: String,
    pub objective_type: String,
    pub status: String,
    pub progress: f32,
    pub confidence: f32,
    pub owner_id: Uuid,
    pub team_id: Option<Uuid>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub team_ids: Option<Vec<Uuid>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrKeyResult {
    pub objective_id: Uuid,
    pub id: Uuid,
    pub title: String,
    pub kr_type: String,
    pub start_value: f32,
    pub target_value: f32,
    pub current_value: f32,
    pub unit: Option<String>,
    pub confidence: f32,
    pub status: String,
    pub owner_id: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    // Enriched
    #[serde(skip_serializing_if = "Option::is_none")]
    pub owner_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub check_ins: Option<Vec<OkrCheckIn>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrCheckIn {
    pub key_result_id: Uuid,
    pub id: Uuid,
    pub previous_value: f32,
    pub new_value: f32,
    pub confidence: f32,
    pub note: Option<String>,
    pub blockers: Option<String>,
    pub author_id: Uuid,
    pub author_name: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OkrAttachment {
    pub entity_id: Uuid,
    pub id: Uuid,
    pub entity_type: String,
    pub workspace_id: Uuid,
    pub file_name: String,
    pub file_size: i64,
    pub mime_type: String,
    pub uploaded_by: Uuid,
    pub uploader_name: String,
    pub created_at: DateTime<Utc>,
}

// ─── Request Types ───

#[derive(Debug, Deserialize)]
pub struct CreateObjectiveRequest {
    pub title: String,
    pub description: Option<String>,
    #[serde(default = "default_objective_type")]
    pub objective_type: String,
    pub cycle: String,
    pub owner_id: Option<Uuid>,
    pub team_id: Option<Uuid>,
    pub team_ids: Option<Vec<Uuid>>,
    pub parent_objective_id: Option<Uuid>,
}

fn default_objective_type() -> String {
    "committed".to_string()
}

#[derive(Debug, Deserialize)]
pub struct UpdateObjectiveRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub objective_type: Option<String>,
    pub cycle: Option<String>,
    pub status: Option<String>,
    pub owner_id: Option<Uuid>,
    pub team_id: Option<Uuid>,
    pub team_ids: Option<Vec<Uuid>>,
    pub parent_objective_id: Option<Uuid>,
}

#[derive(Debug, Deserialize)]
pub struct CreateKeyResultRequest {
    pub title: String,
    #[serde(default = "default_kr_type")]
    pub kr_type: String,
    #[serde(default)]
    pub start_value: f32,
    pub target_value: f32,
    pub unit: Option<String>,
    pub owner_id: Option<Uuid>,
}

fn default_kr_type() -> String {
    "number".to_string()
}

#[derive(Debug, Deserialize)]
pub struct UpdateKeyResultRequest {
    pub title: Option<String>,
    pub kr_type: Option<String>,
    pub start_value: Option<f32>,
    pub target_value: Option<f32>,
    pub current_value: Option<f32>,
    pub unit: Option<String>,
    pub confidence: Option<f32>,
    pub status: Option<String>,
    pub owner_id: Option<Uuid>,
}

#[derive(Debug, Deserialize)]
pub struct CreateCheckInRequest {
    pub new_value: f32,
    pub confidence: f32,
    pub note: Option<String>,
    pub blockers: Option<String>,
}

// ─── Query Params ───

#[derive(Debug, Deserialize)]
pub struct OkrQueryParams {
    pub cycle: Option<String>,
    pub team_id: Option<Uuid>,
}

File: frontend/types/okr.ts
export interface OkrObjective {
  workspace_id: string
  id: string
  title: string
  description: string | null
  objective_type: OkrObjectiveType
  cycle: string
  status: OkrStatus
  owner_id: string
  team_id: string | null
  team_ids?: string[] | null
  parent_objective_id: string | null
  progress: number
  confidence: number
  created_at: string
  updated_at: string
  key_results?: OkrKeyResult[]
  owner_name?: string | null
  team_name?: string | null
  children?: OkrObjectiveChild[]
}

export interface OkrObjectiveChild {
  parent_objective_id: string
  child_objective_id: string
  workspace_id: string
  child_title: string
  child_progress: number
  child_confidence: number
  child_team_id: string | null
}

export interface OkrObjectiveByCycle {
  workspace_id: string
  cycle: string
  objective_id: string
  title: string
  objective_type: OkrObjectiveType
  status: OkrStatus
  progress: number
  confidence: number
  owner_id: string
  team_id: string | null
  team_ids?: string[] | null
  key_results?: OkrKeyResult[]
}

export interface OkrKeyResult {
  objective_id: string
  id: string
  title: string
  kr_type: KrType
  start_value: number
  target_value: number
  current_value: number
  unit: string | null
  confidence: number
  status: OkrStatus
  owner_id: string
  created_at: string
  updated_at: string
  owner_name?: string | null
  check_ins?: OkrCheckIn[]
}

export interface OkrCheckIn {
  key_result_id: string
  id: string
  previous_value: number
  new_value: number
  confidence: number
  note: string | null
  blockers: string | null
  author_id: string
  author_name: string
  created_at: string
}

export interface OkrAttachment {
  entity_id: string
  id: string
  entity_type: string
  workspace_id: string
  file_name: string
  file_size: number
  mime_type: string
  uploaded_by: string
  uploader_name: string
  created_at: string
}

export type OkrObjectiveType = 'committed' | 'aspirational' | 'learning'
export type OkrStatus = 'on_track' | 'at_risk' | 'behind' | 'achieved' | 'canceled'
export type KrType = 'number' | 'percentage' | 'binary'
export const OKR_TYPE_CONFIG: Record<OkrObjectiveType, { label: string; color: string }> = {
  committed: { label: 'Committed', color: '#22c55e' },
  aspirational: { label: 'Aspirational', color: '#f59e0b' },
  learning: { label: 'Learning', color: '#3b82f6' },
}

export const OKR_STATUS_CONFIG: Record<OkrStatus, { label: string; color: string }> = {
  on_track: { label: 'No Prazo', color: '#22c55e' },
  at_risk: { label: 'Em Risco', color: '#f59e0b' },
  behind: { label: 'Atrasado', color: '#ef4444' },
  achieved: { label: 'Alcançado', color: '#8b5cf6' },
  canceled: { label: 'Cancelado', color: '#6b7280' },
}

export function getCurrentCycle(): string {
  const now = new Date()
  const quarter = Math.ceil((now.getMonth() + 1) / 3)
  return `${now.getFullYear()}-Q${quarter}`
}

export function getConfidenceColor(confidence: number): string {
  if (confidence >= 0.7) return '#22c55e'
  if (confidence >= 0.4) return '#f59e0b'
  return '#ef4444'
}

export function getProgressPercent(kr: OkrKeyResult): number {
  const range = kr.target_value - kr.start_value
  if (Math.abs(range) < 0.0001) {
    return kr.kr_type === 'binary' ? (kr.current_value >= 1 ? 100 : 0) : 0
  }
  return Math.min(100, Math.max(0, ((kr.current_value - kr.start_value) / range) * 100))
}
